feat(spark): add Spark 4.1 support

yihua opened the pull request:
### Describe the issue this Pull Request addresses

This PR adds support for Hudi on Spark 4.1, closes #17663.

### Summary and Changelog

Add Spark 4.1 support to Apache Hudi, introducing a new `hudi-spark4.1.x` adapter module that handles API changes between Spark 4.0 and 4.1.

**Package relocations:**
- `StreamExecution` moved from `o.a.s.sql.execution.streaming` to `o.a.s.sql.execution.streaming.runtime`
- `MemoryStream` moved to `o.a.s.sql.execution.streaming.runtime`
- `SerializedOffset` moved to `o.a.s.sql.execution.streaming.runtime`

**API signature changes:**
- `DataSourceV2Relation` extractor: 5 args → 6 args
- `UpdateAction` case class: added extra parameter
- `ParseException` constructor: `stop` parameter removed
- `PartitionedFile`: `locations` parameter removed
- `UnresolvedWith`: added `RECURSIVE` support and `rowLevelLimit`
- `LegacyBehaviorPolicy` config: `String` → `LegacyBehaviorPolicy.Value` for rebase modes

**New types:**
- `GeographyVal` and `GeometryVal` support in `InternalRow` implementations

**Parquet 1.16.0 compatibility (upgraded from 1.15.2 in Spark 4.0):**
- Vectored I/O in `readVectored()` bypasses `FSDataInputStream` virtual methods — fixed `InLineFsDataInputStream` with `OffsetAdjustingStream` wrapper that applies inline offset to all positional reads including vectored I/O used for Parquet log blocks
- Fixed pre-existing bounds check bug in `InLineFsDataInputStream.read()`/`readFully()`: `(length + offset)` → `(position + length)` to correctly detect reads past inline content
- Schema trimming in `HoodieParquetReadSupport.trimParquetType`: exclude non-matching fields from Parquet read schema (Spark schema evolution fills them with null), required by `ParquetRowConverter` assertion in Spark 4.1

**Avro 1.12.1 (upgraded from 1.12.0 in Spark 4.0):**
- `ConvertingGenericData` returns `java.time.Instant`/`LocalDate`/`LocalDateTime` instead of raw `Long`/`Integer` for logical types — added normalization in `HoodieAvroUtils` to convert back to primitives. This needs revisiting (#18606).

**Other:**
- Spark 4.1 Scala version: 2.13.17 (upgraded from 2.13.8)
- Hadoop version: 3.4.2 (upgraded from 3.4.0)
- Added `HoodieSparkUtils.gteqSpark4_1` version helper
- Added `HoodieStreamingSink.QUERY_ID_KEY` constant to avoid cross-version import of relocated `StreamExecution`
- Validation tests for `QUERY_ID_KEY` across all Spark versions (3.3, 3.4, 3.5, 4.0, 4.1)
- CI and Docker bundle validation support for Spark 4.1
- Release script update for Spark 4.1

**Known issues (tracked separately):**
- `TestCreateTable`: skipped on 4.1 only for further investigation (#18609)
- `TestVariantDataType`: JVM SIGSEGV crash on Spark 4.1 (#18605, skipped on 4.1 only)
- Avro 1.12 logical type ClassCastException during record merging with timestamp ordering fields (#18606, tests skipped on 4.1 only)
- Lance tests skipped on Spark 4.1 (#18607)

### Impact

Adds Spark 4.1 as a supported engine version. No changes to existing Spark 3.x or 4.0 behavior except:
- `HoodieParquetReadSupport.trimParquetType`: non-matching fields excluded from Parquet read schema (functionally equivalent — Spark schema evolution fills null instead of parquet-mr)
- `InLineFsDataInputStream`: bounds check fix and vectored I/O support (correctness fix for Parquet log block reads)

### Risk Level

medium

### Documentation Update

Spark 4.1 support is updated in READMEs. I'll update the website separately.

### Contributor's checklist

- [x] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [x] Enough context is provided in the sections above
- [x] Adequate tests were added if applicable

---

### Appendix A: InLineFsDataInputStream fix for Parquet 1.16.0 vectored I/O

#### Background

Hudi stores Parquet data blocks inline within log files for MOR tables. These are read via `InLineFileSystem`, which provides a virtual filesystem view of a slice (at `startOffset` with `length`) within a larger outer file. `InLineFsDataInputStream` translates all read positions by `startOffset` so the reader sees the inline content starting at position 0.

#### Why Parquet 1.16.0 breaks the old code

**Old code path (Spark 3.x / 4.0, Parquet ≤ 1.15.2):**

All reads go through `InLineFsDataInputStream`'s overridden virtual methods:

```
ParquetFileReader.readFully(position, buf, off, len)
  → InLineFsDataInputStream.readFully(position, buf, off, len)   // override
    → outerStream.readFully(startOffset + position, buf, off, len)  // correct ✓
```

The raw inner stream (`this.in`) is never accessed directly by the Parquet reader.

**New code path (Spark 4.1, Parquet 1.16.0, Hadoop 3.4.x):**

Parquet 1.16.0 uses vectored I/O via `readVectored()`. Hadoop's default `FSDataInputStream.readVectored()` bypasses the virtual methods and delegates directly to the wrapped inner stream:

```
ParquetFileReader.readVectored(ranges, allocator)
  → FSDataInputStream.readVectored(ranges, allocator)
    → ((PositionedReadable) this.in).readFully(range.getOffset(), buf)  // direct!
      → rawStream.readFully(offset, buf)   // NO startOffset translation → WRONG DATA
```

The `this.in` field is whatever was passed to `super(wrappedStream)` in the constructor. The old code passed `outerStream.getWrappedStream()` — the raw underlying stream with no offset awareness.

#### The fix

Replace the raw inner stream with an `OffsetAdjustingStream` wrapper:

```java
// Old
super(outerStream.getWrappedStream());  // raw stream, no offset

// New
super(new OffsetAdjustingStream(outerStream, startOffset));  // offset-aware
```

`OffsetAdjustingStream` implements `Seekable` + `PositionedReadable` and adds `startOffset` to every positional read. When Hadoop's `readVectored()` calls `this.in.readFully(offset, ...)`, the offset adjustment is applied correctly:

```
ParquetFileReader.readVectored(ranges, allocator)
  → FSDataInputStream.readVectored(ranges, allocator)
    → ((PositionedReadable) this.in).readFully(range.getOffset(), buf)
      → OffsetAdjustingStream.readFully(position, buf)
        → outer.readFully(startOffset + position, buf)  // correct ✓
```

#### Bounds check fix

Also fixed a pre-existing bug in the positional read bounds check:

```java
// Old — compares buffer layout to file size (semantically wrong)
if ((length + offset) > this.length)

// New — checks if read extends past inline content (correct)
if ((position + length) > this.length)
```

The old check could silently allow reads past the inline content boundary (reading data from adjacent content in the outer file) when `position` is near the end but `offset` is small.

---

### Appendix B: HoodieParquetReadSupport.trimParquetType schema trimming change

#### What changed

In `trimParquetType`, when a nested field exists in the requested schema but NOT in the data file's schema, the behavior changed from keeping the field to excluding it:

```scala
// Old: keep non-matching field in Parquet read schema
Some(field)

// New: exclude non-matching field from Parquet read schema
None
```

#### Why the change is needed

When a nested struct has fields in the requested schema that don't exist in the file (e.g., after schema evolution adds a new column), the old behavior kept these non-matching fields in the Parquet read schema. On Spark 3.x and 4.0, parquet-mr's schema reconciliation handled this gracefully by returning null for missing columns.

On Spark 4.1 (with Parquet 1.16.0), this causes failures because the Parquet read schema contains fields that have no corresponding data in the file. The exact failure path involves Spark's `ParquetRowConverter`, which has an assertion `parquetType.getFieldCount <= catalystType.length` that is violated when the reconciled parquet schema diverges from what the converter expects. This assertion exists since Spark 4.0 but only triggers on Spark 4.1 due to changes in how the Parquet 1.16.0 read path interacts with the converter.

The fix excludes non-matching nested fields from the Parquet read schema and lets Spark's schema evolution layer fill them with null instead.

#### No user-facing impact

The end result is identical for users:

| | Old behavior | New behavior |
|---|---|---|
| **Who provides null for missing field** | parquet-mr (returns null for column not in file) | Spark schema evolution (fills null in output row) |
| **User sees** | null | null |
| **Data correctness** | Same | Same |

The only difference is which layer fills in the null value. This is safe to apply unconditionally across all Spark versions since Spark's schema evolution has always handled missing columns correctly.
---

### Appendix C: Comparison of hudi-spark4.0.x and hudi-spark4.1.x adapter classes

Each Spark 4.0 class has a corresponding Spark 4.1 class. After normalizing version names (`Spark40` → `Spark41`, `4_0` → `4_1`), the differences are:

| Spark 4.0 class | Spark 4.1 class | Diff | Reason |
|---|---|---|---|
| `Spark40HoodieFileScanRDD` | `Spark41HoodieFileScanRDD` | Identical | — |
| `Spark40HoodiePartitionCDCFileGroupMapping` | `Spark41HoodiePartitionCDCFileGroupMapping` | Whitespace only | — |
| `Spark40HoodiePartitionFileSliceMapping` | `Spark41HoodiePartitionFileSliceMapping` | Whitespace only | — |
| `Spark40HoodiePartitionValues` | `Spark41HoodiePartitionValues` | Added `getGeography`/`getGeometry` | New `GeographyVal`/`GeometryVal` types in Spark 4.1 |
| `Spark40HoodieInternalRow` | `Spark41HoodieInternalRow` | Added `getGeography`/`getGeometry` | Same |
| `HoodieSpark40CatalystExpressionUtils` | `HoodieSpark41CatalystExpressionUtils` | Cosmetic (JIRA link, spacing) | — |
| `HoodieSpark40CatalystPlanUtils` | `HoodieSpark41CatalystPlanUtils` | `SerializedOffset` package; `UpdateAction` extra param | Package moved to `runtime`; `UpdateAction` API change |
| `HoodieSpark40SchemaUtils` | `HoodieSpark41SchemaUtils` | Identical | — |
| `Spark4_0Adapter` | `Spark4_1Adapter` | `MemoryStream` package; `DataSourceV2Relation` 6 args; `ParseException` constructor | Package moved to `runtime`; API signature changes |
| `AvroDeserializer` (shared) | `AvroDeserializer` (shared) | `LegacyBehaviorPolicy.Value` for rebase; Avro 1.12.1 `LocalDate`/`Instant` handling | Config API change; Avro 1.12.1 java.time type normalization |
| `AvroSerializer` (shared) | `AvroSerializer` (shared) | `LegacyBehaviorPolicy` config change | Config returns enum directly instead of String |
| `AvroUtils` (shared) | `AvroUtils` (shared) | Identical | — |
| `HoodieSpark4_0AvroDeserializer` | `HoodieSpark4_1AvroDeserializer` | Rebase config: `.toString` removed | `getConf` returns `LegacyBehaviorPolicy.Value` directly |
| `HoodieSpark4_0AvroSerializer` | `HoodieSpark4_1AvroSerializer` | Identical | — |
| `HoodieSpark40PartitionedFileUtils` | `HoodieSpark41PartitionedFileUtils` | `PartitionedFile` no `locations` param; removed `FileStatusWithMetadata` reflection | API simplified in Spark 4.1 |
| `Spark40NestedSchemaPruning` | `Spark41NestedSchemaPruning` | Whitespace only | — |
| `Spark40OrcReader` | `Spark41OrcReader` | Identical | — |
| `Spark40DataSourceUtils` | `Spark41DataSourceUtils` | Identical | — |
| `Spark40LegacyHoodieParquetFileFormat` | `Spark41LegacyHoodieParquetFileFormat` | Large diff (403 lines) | Parquet 1.16.0 vectored I/O; schema trim; `HoodieHadoopStorage`; `Utils` import |
| `Spark40ParquetReader` | `Spark41ParquetReader` | Large diff (332 lines) | Parquet 1.16.0 reader API; `FilterPredicate`/`MessageType` imports; `RebaseDateTime`; `Closeable`/`ZoneId` |
| `Spark40ResolveHudiAlterTableCommand` | `Spark41ResolveHudiAlterTableCommand` | Whitespace only | — |
| `HoodieSpark40Analysis` | `HoodieSpark41Analysis` | `DataSourceV2Relation` 6 args; removed unused imports | API signature change |
| `HoodieSpark4_0ExtendedSqlAstBuilder` | `HoodieSpark4_1ExtendedSqlAstBuilder` | CTE `RECURSIVE`/`rowLevelLimit` support; `UnresolvedWith` extra params | SQL parser enhancements in Spark 4.1 |
| `HoodieSpark4_0ExtendedSqlParser` | `HoodieSpark4_1ExtendedSqlParser` | `stop` param removed from `ParseException` | Constructor change |

yihua committed:
feat(spark): Add Spark 4.1 support

yihua committed:
Fix compilation issue caused by StreamExecution.QUERY_ID_KEY and UpdateAction

yihua committed:
Fix SerializedOffset

yihua committed:
Fix jackson dependencies

yihua committed:
Fix newParseException

yihua committed:
Add support for new types in InternalRow

yihua committed:
Fix compilation issues

yihua committed:
Fix HoodieSpark4_1ExtendedSqlParser

yihua committed:
Fix config for LegacyBehaviorPolicy.Value

yihua committed:
Fix Spark41OrcReader

yihua committed:
Fix serde and Spark41ParquetReader

yihua committed:
Fix Spark41LegacyHoodieParquetFileFormat

yihua committed:
Fix asScala and withCTE

yihua committed:
Fix SqlBase with parameter_substitution_enabled

yihua committed:
Fix scalastyle

yihua committed:
Fix MemoryStream usage

yihua committed:
Fix Spark 4.0 build

yihua committed:
Fix Spark 3.5 profile

yihua committed:
Fix spark4 common POM

yihua committed:
Hook up Spark41 adapter and add more fixes

yihua committed:
Add Spark 4.1 tests in GH CI

yihua committed:
Fix jackson

yihua committed:
Fix lance and parquet-hadoop-bundle conflict

yihua committed:
Fix build wrt lance file format

yihua committed:
Fix InLineFsDataInputStream for parquet 1.16.0 vectored IO compatibility

Parquet-hadoop 1.16.0 (used by Spark 4.1) changed HadoopInputFile.newStream()
to use FileSystem.openFile(), and its default readVectored() path in Hadoop 3.3+
delegates directly to the wrapped PositionedReadable inner stream rather than
through the FSDataInputStream virtual methods. This bypassed InLineFsDataInputStream's
offset translation, causing reads to start from position 0 of the outer file
instead of the inline content's start offset, producing corrupted parquet page
headers (Required field 'uncompressed_page_size' was not found).

Fix by wrapping the inner stream in an OffsetAdjustingStream that implements
Seekable + PositionedReadable and shifts all positional reads by startOffset.
Since VectoredReadUtils.readRangeFrom() calls readFully(range.getOffset(), ...)
on whatever PositionedReadable is in the 'in' field, this ensures vectored reads
also land at the correct absolute position in the outer file.

Also fix the pre-existing bounds check bug in the positional read overrides:
- was: (length + offset) > this.length  (compares buffer slice bounds, not file bounds)
- now: (position + length) > this.length (correctly detects reads past inline content)

yihua committed:
Add Vector and Variant support to Spark 4.1 AvroSerializer/Deserializer

Port the Vector (FLOAT/DOUBLE/INT8 embedding) and Variant type handling from
the Spark 4.0 AvroSerializer/AvroDeserializer to the Spark 4.1 versions.

- AvroSerializer: Add (ArrayType(elementType, false), FIXED) case for VectorLogicalType
  and (VariantType, RECORD) case for variant logicalType
- AvroDeserializer: Add (FIXED, ArrayType(elementType, false)) case for VectorLogicalType
  and (RECORD, VariantType) case for variant logicalType

This fixes TestAvroSerDe test failures (testVectorDoubleSerDe, testVectorFloatByteOrder,
testVariantSerDe, etc.) and the "Test Table with Variant Data Type" CI failure.

yihua committed:
Add Spark 4.1 support to bundle validation CI scripts

Add spark4.1.0 (IMAGE_TAG=flink1200hive313spark410scala213) to ci_run.sh
and run_docker_java17.sh so that validate-bundle-spark4 and docker-java17-test
CI jobs no longer fail with 'invalid tag' due to empty IMAGE_TAG.

Also update validate.sh to handle spark-4.1.x similarly to spark-4.0.x
(Java 17 runtime change, HiveQL non-fatal failure, CLI bundle inclusion).

Add base/build_flink1200hive313spark410scala213.sh for building the Docker
base image when needed.

yihua committed:
Fix Avro 1.12.1 LocalDate cast in Spark 4.1 AvroDeserializer

Avro 1.12.1 (used by Spark 4.1 / Hadoop 3.4 profile) applies DateConversion
via ConvertingGenericData, returning java.time.LocalDate for INT fields with
date logical type. The (INT, IntegerType) and (INT, DateType) cases were
assuming Int and failing with ClassCastException.

Handle LocalDate explicitly by converting via toEpochDay in both cases.

yihua committed:
Fix Spark 4.1 PushVariantIntoScan variant struct in HoodieSparkSchemaConverters

In Spark 4.1, the PushVariantIntoScan optimizer rule rewrites VariantType
columns in the scan requiredSchema as a StructType with positional field
names ("0", "1", ...) annotated with VariantMetadata. This caused
HoodieSparkSchemaConverters.toHoodieType to hit the generic StructType
handler and attempt to create an Avro record with digit-only field names,
resulting in HoodieSchemaException: Illegal initial character: 0.

Fix: add isPushDownVariantStruct to SparkAdapter (default false) with an
override in Spark4_1Adapter that delegates to VariantMetadata.isVariantStruct.
In toHoodieType, detect this case before the generic StructType branch and
convert it to HoodieSchemaType.VARIANT directly.

yihua committed:
Add BLOB type support to Spark 4.1 SQL parser

HoodieSpark4_1ExtendedSqlAstBuilder was missing the blob type case in
visitPrimitiveDataType and the addMetadataForType logic in visitColType.
This caused TestCreateTable BLOB tests to fail in Spark 4.1 with
ParseException: DataType blob is not supported.

yihua committed:
Fix Scalastyle import order in Spark4_1Adapter

Move VariantMetadata import to correct alphabetical position before parser imports.

yihua committed:
Fix Scalastyle import order: VariantMetadata before datasources._

VariantMetadata (capital V, ASCII 86) sorts before datasources._
(underscore, ASCII 95) in case-sensitive lexicographic order, so it
must be placed before the wildcard datasources._ import.

yihua committed:
Remove redundant VariantMetadata import covered by datasources._ wildcard

yihua committed:
Fix Spark 4.1 test failures: BLOB parser, Avro 1.12.1 types, Parquet schema trim

1. Add missing " blob" check to isHoodieCommand() in
   HoodieSpark4_1ExtendedSqlParser so CREATE TABLE with BLOB
   columns routes through Hudi's parser where BLOB is supported.

2. Handle Avro 1.12.1 returning java.time.Instant for timestamp
   fields and java.time.LocalDate for date fields in both the
   Spark 4.1 AvroDeserializer and the shared HoodieAvroUtils
   rewritePrimaryType methods.

3. Fix trimParquetType to exclude fields from the requested schema
   that don't exist in the file schema. Including them violated
   Spark 4.1's stricter ParquetRowConverter assertion
   (parquetType.getFieldCount <= catalystType.length).

yihua committed:
Normalize Avro 1.12.1 logical type values in HoodieAvroUtils

Avro 1.12.1's ConvertingGenericData returns java.time.Instant for
timestamp fields and java.time.LocalDate for date fields. When these
values are used as ordering fields in record merging, they cause
ClassCastException because one side has Instant while the other has
Long. Normalize these values back to primitives (Long/Integer) in
convertValueForSpecificDataTypes to ensure consistent types.

yihua committed:
Fix BLOB metadata in nested struct fields for Spark 4.1

visitComplexColType in HoodieSpark4_1ExtendedSqlAstBuilder was missing
the addMetadataForType call that exists in the Spark 4.0 version.
This caused BLOB columns inside STRUCT<...> to lose their type
metadata annotation.

CTTY commented on hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/HoodieParquetReadSupport.scala:
```diff
@@ -134,7 +134,11 @@ object HoodieParquetReadSupport {
+              // Exclude it from the Parquet read schema; Spark's schema evolution
+              // will fill it with null. Including it would violate Spark 4.1's
+              // ParquetRowConverter assertion (parquetType.getFieldCount <= catalystType.length).
+              None
```
Is returning None the correct behavior for previous Spark versions?

yihua replied:
I'll revisit this.  This is for schema evolution for nested columns, and the case where the field exists in the requested schema but not in the file.  So, making it `None` will keep the values as `null`, the same behavior as before.

CTTY commented on hudi-spark-datasource/hudi-spark3.3.x/src/main/scala/org/apache/spark/sql/HoodieSpark33CatalystPlanUtils.scala:
```diff
@@ -133,4 +134,18 @@ object HoodieSpark33CatalystPlanUtils extends BaseHoodieCatalystPlanUtils {
     }
   }
+
+  override def unapplyUpdateAction(mergeAction: Any): Option[(Option[Expression], Seq[Assignment])] = {
```
nit: maybe add a default implementation? so we don't have to change every previous utils

yihua replied:
Unfortunately, that's not possible because the number of arguments for `UpdateAction` has changed, so we cannot have a default implementation to share.

yihua committed:
Fix ShowIndexes -> HoodieShowIndexes in Spark 4.1 adapter

The lance-spark-base dependency provides its own ShowIndexes class with
only 1 parameter (table), which shadows the Hudi 2-parameter version.
Use HoodieShowIndexes consistently, matching the Spark 4.0 adapter.

yihua committed:
Add missing injectScalarFunctions and injectPlannerStrategies to Spark4_1Adapter

These methods were added to SparkAdapter upstream but not implemented
in the Spark 4.1 adapter, causing an abstract class error.

yihua committed:
Upgrade Spark 4.1 version from 4.1.0 to 4.1.1

Update all references: pom.xml version property, CI workflow matrix,
bundle validation scripts, and Docker build script (renamed from
spark410 to spark411 in image tag).

yihua committed:
Disable TestHoodieSparkSchemaUtils due to pre-existing failures

This test fails on all Spark versions (3.3, 3.4, 3.5, 4.0, 4.1) with
HoodieSchemaException in convertStructTypeToHoodieSchema, indicating a
pre-existing issue on upstream master unrelated to Spark 4.1 support.

yihua committed:
Fix Scalastyle import order in TestHoodieSparkSchemaUtils

Split grouped import {Disabled, Test} into separate imports to satisfy
Scalastyle ImportOrderChecker which treats grouped imports as a package
prefix, breaking alphabetical ordering relative to Assertions.

yihua committed:
Disable TestInLineFileSystem due to pre-existing failures on master

yihua committed:
Disable TestInsertTable3 due to pre-existing bucket index bulk insert failures

The bucket index tests fail with HoodieNotSupportedException on both
Spark 4.0 and 4.1 when Spark native row writer is disabled. This is a
pre-existing issue on upstream master, not related to Spark 4.1 support.

yihua committed:
Disable TestExpressionIndex and TestSecondaryIndex on Spark 4.1

Both tests fail with expression/secondary index bootstrap errors and
FAILED_READ_FILE parquet errors on Spark 4.1. These are Spark 4.x
compatibility issues to be investigated separately.

yihua committed:
Switch @Ignore to @DoNotDiscover for disabled ScalaTest suites

@Ignore still discovers and instantiates suites, running beforeAll/afterAll
which can fail. @DoNotDiscover prevents discovery entirely.

yihua committed:
Disable TestHoodieParquetReadSupport, TestInsertTable, TestInsertTable5

TestHoodieParquetReadSupport: pre-existing testSchemaTrimming failure.
TestInsertTable and TestInsertTable5: Spark 4.1 parquet read failures
in DML tests with task stage failures.

yihua committed:
Make failing ScalaTest suites abstract to prevent discovery

@DoNotDiscover and @Ignore do not prevent scalatest-maven-plugin
wildcardSuites from discovering and running these suites. Making the
classes abstract ensures they cannot be instantiated by any runner.

yihua committed:
Disable TestInsertTableWithPartitionBucketIndex and TestCreateTable on Spark 4.1

TestInsertTableWithPartitionBucketIndex: fails with HoodieException
"Failed to get hashing config instant to load" in partition bucket
index initialization.

TestCreateTable: Spark 4.1 parser rejects type aliases (DOUBLE, FLOAT,
INT8, BOOLEAN) that worked in earlier Spark versions.

yihua committed:
Bump test JVM heap from 3g to 4g globally

Spark tests with reuseForks=true accumulate heap pressure across
sequential test suites. The 3g limit causes intermittent OOM failures
in heavy test jobs like java-tests-part2 (TestCOWDataSource +
TestMORDataSource) and scala-dml-tests. Bump to 4g across all
profiles (default, java11, java17, spark4.1) for stability.

yihua committed:
Re-enable locally passing tests and switch DML CI to Corretto JDK 17

Re-enable TestInsertTable, TestInsertTable3, TestInsertTable5,
TestExpressionIndex, TestSecondaryIndex — all pass locally on Spark 4.1.

Switch test-spark-java17-scala-dml-tests CI job from Temurin to
Corretto JDK 17 to avoid SIGSEGV JVM crash in
StubRoutines::jshort_disjoint_arraycopy (Temurin 17.0.18+8 bug).

yihua committed:
Refine test disabling: use @DisabledOnSpark4, re-enable passing tests

- TestHoodieParquetReadSupport: @Disabled -> @DisabledOnSpark4 (still
  runs on Spark 3.5, skipped on Spark 4.x)
- TestHoodieSparkSchemaUtils: @Disabled -> @DisabledOnSpark4 (same)
- TestInLineFileSystem: restore @Disabled (pre-existing master issue)
- TestInsertTableWithPartitionBucketIndex: re-enable (passes on 4.1)
- TestVariantDataType: skip on Spark 4.x via cancel() in beforeAll
  (JVM SIGSEGV crash), still runs on Spark 3.5

yihua committed:
Disable TestCreateTable on Spark 4.x via cancel() in beforeAll

TestCreateTable hangs on inline compaction during MOR table creation
on Spark 4.1. Use cancel() so it still runs on Spark 3.5.

yihua committed:
Only disable failing tests on Spark 4.1, re-enable on all other versions

Add HoodieSparkUtils.gteqSpark4_1 helper. Switch all test disabling
from blanket @Disabled/@DisabledOnSpark4 to Spark 4.1-only skips:
- TestHoodieParquetReadSupport: Assumptions.assumeFalse on Spark 4.1
- TestHoodieSparkSchemaUtils: @BeforeEach skip on Spark 4.1
- TestCreateTable: cancel() on Spark 4.1 (inline compaction hang)
- TestVariantDataType: cancel() on Spark 4.1 (JVM SIGSEGV crash)
- TestInLineFileSystem: re-enabled (CI runs on Spark 3.5)

yihua committed:
Restore @Disabled for tests that fail on all Spark versions

TestInLineFileSystem, TestHoodieSparkSchemaUtils, and
TestHoodieParquetReadSupport fail on Spark 3.5 and 4.0 too (not just
4.1). These are pre-existing failures on master — use @Disabled
globally until the underlying issues are fixed.

yihua committed:
Fix test disabling: use withFixture for Spark 4.1 skips, restore @Disabled for pre-existing failures

- TestCreateTable, TestVariantDataType: switch from cancel() in
  beforeAll (causes suite abort = failure) to withFixture returning
  Canceled (properly marks tests as skipped)
- TestInsertTable4: skip StageParallelismListener tests on Spark 4.1
  via assume() (numTasks assertion differs)
- TestInLineFileSystem, TestHoodieSparkSchemaUtils,
  TestHoodieParquetReadSupport: restore @Disabled (fail on all Spark
  versions including 3.5, pre-existing on master)

yihua committed:
Revert StageParallelismListener skip in TestInsertTable4

yihua committed:
Fix Scala 2.12 compilation: use fully qualified Outcome and Canceled

Outcome and Canceled are not automatically in scope on Scala 2.12.
Use org.scalatest.Outcome and org.scalatest.Canceled explicitly.

yihua committed:
Use proper imports for Outcome and Canceled in Scala 2.12

Import org.scalatest.{Canceled, Outcome} instead of using fully
qualified names.

CTTY commented on hudi-client/hudi-spark-client/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/TestHoodieParquetReadSupport.scala:
```diff
@@ -19,8 +19,9 @@ package org.apache.spark.sql.execution.datasources.parquet
-import org.junit.jupiter.api.{Assertions, Test}
+import org.junit.jupiter.api.{Assertions, Disabled, Test}
 
+@Disabled("Pre-existing failure on master: testSchemaTrimming_atLeastOneFieldMatches")
```
should we just disable the test not the entire suite?

yihua replied:
This is removed now.

yihua committed:
Re-enable TestInLineFileSystem and TestHoodieParquetReadSupport

CTTY commented on hudi-hadoop-common/src/test/java/org/apache/hudi/common/fs/inline/TestInLineFileSystem.java:
```diff
@@ -51,6 +52,8 @@
  * Tests {@link InLineFileSystem}.
  */
+
+@Disabled("Pre-existing failure on master: testFileSystemApis")
```
Same here, should we only disable the unit test?

yihua replied:
This is fixed now.

CTTY commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala:
```diff
@@ -57,9 +57,11 @@ object HoodieAnalysis extends SparkAdapterSupport {
+    } else if (HoodieSparkUtils.isSpark4_0) {
       "org.apache.spark.sql.hudi.analysis.HoodieSpark40DataSourceV2ToV1Fallback"
-    else if (HoodieSparkUtils.isSpark3_5)
+    } else if (HoodieSparkUtils.isSpark3_5)
```
nit: no brackets for spark 3_5 and 3_4

yihua replied:
fixed

CTTY commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/dml/insert/TestInsertTable.scala:
```diff
@@ -28,7 +28,11 @@ import org.apache.hudi.testutils.HoodieClientTestUtils.createMetaClient
 
+
+// @DoNotDiscover
+// class TestInsertTable extends HoodieSparkSqlTestBase {
```
Unneeded comment

yihua replied:
fixed

CTTY commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/dml/insert/TestInsertTable3.scala:
```diff
@@ -27,6 +27,8 @@ import org.apache.hudi.execution.bulkinsert.BulkInsertSortMode
 import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase
 import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase.getLastCommitMetadata
 import org.junit.jupiter.api.Assertions.assertEquals
+import org.scalatest.DoNotDiscover
```
unneeded import

yihua replied:
fixed

CTTY commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/dml/insert/TestInsertTable5.scala:
```diff
@@ -23,6 +23,8 @@ import org.apache.hudi.DataSourceWriteOptions.{DROP_INSERT_DUP_POLICY, FAIL_INSE
 import org.apache.hudi.HoodieSparkUtils
 
 import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase
+import org.scalatest.DoNotDiscover
```
unneeded import

yihua replied:
fixed

CTTY commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/feature/index/TestExpressionIndex.scala:
```diff
@@ -59,11 +59,13 @@ import org.apache.spark.sql.hudi.command.{CreateIndexCommand, ShowIndexesCommand
 import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase
 import org.apache.spark.sql.types._
 import org.junit.jupiter.api.Assertions.{assertEquals, assertFalse, assertTrue}
+import org.scalatest.DoNotDiscover
```
Unneeded import

yihua replied:
fixed

CTTY commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/feature/index/TestSecondaryIndex.scala:
```diff
@@ -37,11 +37,13 @@ import org.apache.hudi.storage.StoragePath
 import org.apache.spark.sql.SaveMode
 import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase
 import org.junit.jupiter.api.Assertions.{assertEquals, assertFalse, assertNotNull, assertTrue}
+import org.scalatest.DoNotDiscover
```
unneeded import

yihua replied:
fixed

yihua committed:
Address PR review: add brackets for consistency, remove unused imports/comments

- HoodieAnalysis: add brackets to isSpark3_5/isSpark3_4 branches
- Remove unused DoNotDiscover imports and commented-out code from
  TestInsertTable, TestInsertTable3, TestInsertTable5,
  TestExpressionIndex, TestSecondaryIndex

CTTY reviewed (approved):
LGTM, thanks for working on this!

yihua reviewed (commented):
@hudi-agent review once

yihua committed:
Fix TestInLineFileSystem for corrected bounds check in InLineFsDataInputStream

The bounds check in read()/readFully() was fixed from (length + offset)
to (position + length) to correctly detect reads past inline content.
The old check coincidentally caught buffer-overflow cases (offset >
buffer size) as IOException. With the corrected check, buffer-overflow
cases now propagate as IndexOutOfBoundsException from Hadoop's
FSInputStream.validatePositionedReadArgs. Update test expectations
accordingly and add explicit inline-content-overflow assertions.

hudi-agent commented on hudi-spark-datasource/hudi-spark4.1.x/src/main/scala/org/apache/hudi/client/model/Spark41HoodieInternalRow.scala:
```diff
@@ -0,0 +1,52 @@
+  override def getGeography(ordinal: Int): GeographyVal = {
+    ruleOutMetaFieldsAccess(ordinal, classOf[VariantVal])
+    sourceRow.getGeography(rebaseOrdinal(ordinal))
+  }
```
🤖 Looks like a copy/paste from `getVariant` — `getGeography` and `getGeometry` (line 38 and 43) both pass `classOf[VariantVal]` to `ruleOutMetaFieldsAccess`, but they should pass `classOf[GeographyVal]` and `classOf[GeometryVal]` respectively. Otherwise the type-mismatch error message will misreport the accessed type as `VariantVal`.

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
Fixed.

hudi-agent commented on hudi-spark-datasource/hudi-spark4.1.x/src/main/scala/org/apache/spark/sql/adapter/Spark4_1Adapter.scala:
```diff
@@ -0,0 +1,252 @@
+  override def createAvroSerializer(rootCatalystType: DataType, rootType: HoodieSchema, nullable: Boolean): HoodieAvroSerializer =
+    new HoodieSpark4_1AvroSerializer(rootCatalystType, rootType.toAvroSchema, nullable)
+
+  override def createAvroDeserializer(rootType: HoodieSchema, rootCatalystType: DataType): HoodieAvroDeserializer =
```
🤖 The `stop: Origin` parameter is dropped here vs. the Spark 4.0 implementation (`new ParseException(command, start, stop, errorClass, params)`). Is this intentional because Spark 4.1's `ParseException` constructor signature changed, or did `stop` get accidentally lost? If `stop` is still available in 4.1, losing it will degrade error position reporting.

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
Spark 4.1 removes `stop` as an argument.

hudi-agent commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -1222,6 +1247,29 @@ private static boolean needsRewriteToString(Schema schema, boolean isEnum) {
+      java.time.Instant instant = (java.time.Instant) value;
+      return instant.getEpochSecond() * 1000000L + instant.getNano() / 1000L;
+    }
+    return (Long) value;
```
🤖 `avroValueToLong` unconditionally converts an `Instant` to microseconds, but it's called from the `TimestampMillis -> TimestampMicros` branch in `rewritePrimaryType`: `DateTimeUtils.millisToMicros(avroValueToLong(oldValue))`. If Avro 1.12.1 returns an `Instant` for a `TimestampMillis` field, this returns micros, and `millisToMicros` then multiplies by 1000 again — off by 1000x. Could you confirm Avro 1.12.1 only does this conversion for `TimestampMicros`, or should the helper take the source logical type into account? @yihua

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
#18606 to follow up.

hudi-agent commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -660,7 +660,28 @@ public static Object convertValueForSpecificDataTypes(Schema fieldSchema,
+    if (value instanceof java.time.Instant) {
+      java.time.Instant instant = (java.time.Instant) value;
+      return instant.getEpochSecond() * 1000000L + instant.getNano() / 1000L;
+    } else if (value instanceof java.time.LocalDate) {
```
🤖 For a `LocalTimestampMillis`/`LocalTimestampMicros` source, `LocalDateTime.toInstant(ZoneOffset.UTC)` reinterprets the wall-clock time as UTC. That matches Avro's local-timestamp encoding (epoch as if UTC), but only when the LocalDateTime came from a local-timestamp logical type. If Avro 1.12.1 ever returns a `LocalDateTime` for a regular `TimestampMillis`/`TimestampMicros`, this would silently produce a different value than the original raw long. Worth a quick check that this only triggers for the local-timestamp variants.

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
This is removed. #18606 to follow up.

hudi-agent commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/ddl/TestCreateTable.scala:
```diff
@@ -42,11 +42,21 @@ import org.apache.spark.sql.hudi.common.HoodieSparkSqlTestBase
 
+  // TODO(SPARK-4.1): Re-enable after fixing inline compaction hang on Spark 4.1
+  override def withFixture(test: NoArgTest): Outcome = {
+    if (HoodieSparkUtils.gteqSpark4_1) {
```
🤖 Disabling all of `TestCreateTable` for Spark 4.1 due to an inline compaction hang during MOR creation feels like a fairly significant regression to mask. Could you open a tracked issue (referenced in the TODO) and consider whether the hang itself is a blocker for the Spark 4.1 release vs. shipping with the tests skipped? Same question for the SIGSEGV in `TestVariantDataType`. @yihua

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
Follow-up:  #18609

hudi-agent commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/streaming/HoodieSourceOffset.scala:
```diff
@@ -64,8 +65,14 @@ object HoodieSourceOffset {
+        sparkAdapter.getCatalystPlanUtils.extractJsonFromSerializedOffset(offset) match {
+          case Some(json) => fromJson(json)
+          case None => throw new IllegalArgumentException(s"Unsupported offset type: ${offset.getClass}")
+        }
```
🤖 Previously a non-`SerializedOffset`, non-`HoodieSourceOffset` value would surface as a `MatchError`; now it surfaces as an `IllegalArgumentException` only if the adapter doesn't recognize it. Have you verified that the Spark 3.x `extractJsonFromSerializedOffset` implementations match every offset type Spark could pass in here, including any `MicroBatchOffset` or future variant? Otherwise behavior diverges silently from older versions.

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
The behavior is equivalent (`MatchError` → `IllegalArgumentException` for unrecognized types). `SerializedOffset` is the only non-Hudi offset type that reaches this code path in Spark's streaming framework.

hudi-agent commented on hudi-spark-datasource/hudi-spark4.1.x/src/main/scala/org/apache/spark/sql/adapter/Spark4_1Adapter.scala:
```diff
@@ -0,0 +1,252 @@
+
+/**
+ * Implementation of [[SparkAdapter]] for Spark 4.0.x branch
+ */
```
🤖 nit: the class Javadoc still says "Spark 4.0.x branch" — looks like a copy-paste from `Spark4_0Adapter`; could you update it to "Spark 4.1.x branch"?

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
Fixed

hudi-agent commented on hudi-spark-datasource/hudi-spark4.1.x/src/main/scala/org/apache/spark/sql/adapter/Spark4_1Adapter.scala:
```diff
@@ -0,0 +1,252 @@
+
+  /**
+   * TODO
+   *
```
🤖 nit: the `TODO` placeholder was never replaced — `createParquetFileReader` right above has a complete Javadoc; it might be worth filling this in similarly (or at minimum dropping the empty `@param` stubs).

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
Will address this nit separately as this is from existing code.

yihua replied:
Fixed

hudi-agent commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieSparkBaseAnalysis.scala:
```diff
@@ -191,12 +191,14 @@ case class ResolveReferences(spark: SparkSession) extends Rule[LogicalPlan]
                 resolveExpressionByPlanChildren(_, m))
               DeleteAction(resolvedDeleteCondition)
-            case UpdateAction(updateCondition, assignments) =>
+            case action if catalystPlanUtils.unapplyUpdateAction(action).isDefined =>
```
🤖 nit: `unapplyUpdateAction` is called twice here — once for the `.isDefined` guard and once for `.get`. Could you capture the result in a local `val` (or restructure as a nested `match`) to avoid the double invocation?

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
We can fix this.

yihua replied:
This has to be like this because of Scala's guard pattern.  `unapplyUpdateAction` overhead is negligible, so it is fine.

hudi-agent commented on hudi-spark-datasource/hudi-spark4.1.x/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/Spark41LegacyHoodieParquetFileFormat.scala:
```diff
@@ -0,0 +1,474 @@
+/**
+ * This class is an extension of [[ParquetFileFormat]] overriding Spark-specific behavior
+ * that's not possible to customize in any other way
+ *
```
🤖 nit: the Javadoc says "version of [[AvroDeserializer]] impl" but this class extends `ParquetFileFormat`, not `AvroDeserializer` — could you update the reference to `[[ParquetFileFormat]]`?

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
Ignore for now as this is copied from previous Spark version support.

hudi-agent commented on hudi-spark-datasource/hudi-spark4.1.x/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/Spark41LegacyHoodieParquetFileFormat.scala:
```diff
@@ -0,0 +1,474 @@
+        SQLConf.LEGACY_PARQUET_NANOS_AS_LONG.key,
+        SQLConf.LEGACY_PARQUET_NANOS_AS_LONG.defaultValueString).toBoolean
+    )
+    hadoopConf.setBoolean(SQLConf.PARQUET_INFER_TIMESTAMP_NTZ_ENABLED.key, sqlConf.parquetInferTimestampNTZEnabled)
```
🤖 nit: `LEGACY_PARQUET_NANOS_AS_LONG` is already set on lines 113–118 via the string-based approach (with a comment explaining why). This second `setBoolean` silently overwrites it with the direct accessor — could you remove the duplicate, keeping only whichever approach is intentional?

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
Existing code from Spark. Ignore for now.

hudi-agent commented on hudi-spark-datasource/hudi-spark4.1.x/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/Spark41LegacyHoodieParquetFileFormat.scala:
```diff
@@ -0,0 +1,474 @@
+    val timeZoneId = Option(sqlConf.sessionLocalTimeZone)
+    // Should always be set by FileSourceScanExec creating this.
+    // Check conf before checking option, to allow working around an issue by changing conf.
+    val returningBatch = sqlConf.parquetVectorizedReaderEnabled &&
```
🤖 nit: `supportsColumnar(...)` already returns `Boolean`, so `.toString.equals("true")` is a no-op round-trip — could you simplify to just `&& supportsColumnar(sparkSession, resultSchema)`?

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

yihua replied:
Existing code from Spark.  Ignore for now.

hudi-agent reviewed (commented):
🤖 *This review was generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.*

Thanks for working on this! This PR adds Spark 4.1 support by introducing a new `hudi-spark4.1.x` module that parallels the existing 4.0 implementation, refactors shared Spark 4 code, and threads the necessary adapter, parser, and reader changes through `hudi-spark-common` and `hudi-common` (including Avro 1.12.1 timestamp normalization and an `OffsetAdjustingStream` fix for vectored reads). A few correctness items worth a closer look — notably the `Spark41HoodieInternalRow` geography/geometry accessors, a dropped `stop: Origin` argument in `Spark4_1Adapter`, the new `avroValueToLong` timestamp conversion paths, and the `TestCreateTable` suite being disabled wholesale for Spark 4.1. Please take a look at the inline comments, and this should be ready for a Hudi committer or PMC member to take it from here. A few minor copy-paste oversights (wrong version string in the `Spark4_1Adapter` class comment and a forgotten `TODO` Javadoc) plus one double-invocation in the merge-action analysis worth a quick cleanup. a few nits in `Spark41LegacyHoodieParquetFileFormat`; the other new files look clean.

yihua committed:
Exclude non-matching fields from Parquet read schema unconditionally

When trimming the Parquet read schema against the file schema, fields
that exist in the requested schema but not in the file were previously
kept (Some(field)), relying on parquet-mr to return null for missing
columns. This is functionally equivalent to excluding them (None) and
letting Spark's schema evolution fill null, but the latter is more
correct and avoids violating ParquetRowConverter's assertion in Spark
4.1 (parquetType.getFieldCount <= catalystType.length).

Apply unconditionally since the end result is the same: null for the
missing field regardless of which layer provides it.

yihua commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -660,7 +660,28 @@ public static Object convertValueForSpecificDataTypes(Schema fieldSchema,
+   * Avro 1.12.1 automatically converts logical types to java.time objects, but Hudi expects
+   * the raw Long/Integer values internally.
+   */
+  private static Object normalizeAvroLogicalTypeValue(Object value) {
```
Need to revisit this change.

yihua replied:
Removed for now. Follow-up: #18606

yihua committed:
Minor changes

yihua commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieStreamingSink.scala:
```diff
@@ -323,7 +323,11 @@ class HoodieStreamingSink(sqlContext: SQLContext,
+  // Spark 4.1 moves the StreamExecution class from `org.apache.spark.sql.execution.streaming`
+  // to `org.apache.spark.sql.execution.streaming.runtime`, so to maintain compatibility with
+  // different Spark versions, we define the key here
+  val QUERY_ID_KEY = "sql.streaming.queryId"
```
nit: Good to add validation in a test that `StreamExecution.QUERY_ID_KEY` has same String value as `HoodieStreamingSink.QUERY_ID_KEY`.

yihua replied:
Fixed

yihua committed:
Add test to validate HoodieStreamingSink.QUERY_ID_KEY across Spark versions

Spark 4.1 moved StreamExecution to a different package. We maintain a
local copy of QUERY_ID_KEY in HoodieStreamingSink. These tests ensure
the constant stays in sync with the actual Spark value for each version:
- Spark 3.3/3.4/3.5/4.0: imports from o.a.s.sql.execution.streaming
- Spark 4.1: imports from o.a.s.sql.execution.streaming.runtime

yihua committed:
Minor change in TestInLineFileSystem.java

yihua committed:
Remove PushVariantIntoScan workaround and isPushDownVariantStruct

Remove the isPushDownVariantStruct check from HoodieSparkSchemaConverters
and the corresponding SparkAdapter method and Spark4_1Adapter override.
This workaround handled Spark 4.1's PushVariantIntoScan optimizer rewriting
VariantType as StructType with digit-only field names. Needs a proper fix
before re-enabling TestVariantDataType on Spark 4.1.

yihua committed:
Minor changes

yihua committed:
Disable tests affected by Avro 1.12 logical type ClassCastException on Spark 4.1

Skip specific tests on Spark 4.1 that fail due to ClassCastException
when reading ordering fields with logical types (Long vs Instant,
Integer vs LocalDate). Avro 1.12's ConvertingGenericData returns
java.time objects instead of raw primitives, causing type mismatches
during record merging and schema evolution.

- TestMergeIntoTable: "Test Different Type of PreCombineField"
- TestSpark3DDL: "Test alter column types", "Test alter column types 2"

Tracked in https://github.com/apache/hudi/issues/18606

yihua committed:
Remove redundant argLine from spark4.1 profile (already set by java17 profile)

yihua committed:
Add issue reference for skipping lance tests on Spark 4.1

yihua committed:
Revert changes in HoodieAvroUtils

yihua committed:
Add blob, variant, vector support and Databricks Runtime fix on Spark 4.1 to align with Spark 4.0 implementation

yihua committed:
Fix Scalastyle import order in HoodieSpark4_1ExtendedSqlAstBuilder

yihua committed:
Address PR review comments and update docs for Spark 4.1

- Fix Spark41HoodieInternalRow: getGeography/getGeometry passed wrong
  class (VariantVal) to ruleOutMetaFieldsAccess, now uses correct
  GeographyVal/GeometryVal
- Fix Spark4_1Adapter Javadoc: "Spark 4.0.x" -> "Spark 4.1.x"
- Fix createOrcFileReader Javadoc: replace TODO with proper docs
- Update README.md and hudi-spark-datasource/README.md with Spark 4.1
  build profile and module

yihua committed:
Update release scripts for Spark 4.1 bundle
